server/vertx/src/main/scala/sttp/tapir/server/vertx/decoders/VertxInputDecoders.scala

```diff
@@ -0,0 +1,128 @@
+   * @param endpointOptions endpoint options (execution context, etc.)
+   * @param ect the error class to convert to, if the user recovers error
+   * @tparam E Error parameter type, if it should be caught and handled by the user
+   */
+  private [vertx] def decodeBodyAndInputsThen[E](
+    endpoint: Endpoint[_, E, _, _],
+    rc: RoutingContext,
+    logicHandler: Params => Unit,
+  )(implicit endpointOptions: VertxEndpointOptions, ect: Option[ClassTag[E]]): Unit = {
+    decodeBodyAndInputs(endpoint, rc) match {
+      case values: DecodeInputsResult.Values =>
+        InputValues(endpoint.input, values) match {
+          case InputValuesResult.Value(params, _) =>
+            logicHandler(params)
+          case InputValuesResult.Failure(_, failure) =>
+            tryEncodeError(endpoint, rc, failure)
+        }
+      case DecodeInputsResult.Failure(input, failure) =>
+        val decodeFailureCtx = DecodeFailureContext(input, failure)
+        endpointOptions.decodeFailureHandler(decodeFailureCtx) match {
+          case DecodeFailureHandling.NoMatch =>
+            endpointOptions.logRequestHandling.decodeFailureNotHandled(endpoint, decodeFailureCtx)(endpointOptions.logger)
+            rc.response.setStatusCode(404).end()
+          case DecodeFailureHandling.RespondWithResponse(output, value) =>
+            endpointOptions.logRequestHandling.decodeFailureHandled(endpoint, decodeFailureCtx, value)(endpointOptions.logger)
+            VertxOutputEncoders.apply(output, value)(endpointOptions)(rc)
+        }
+
+    }
+  }
+
+  private def decodeBodyAndInputs(e: Endpoint[_, _, _, _], rc: RoutingContext)
+                                         (implicit serverOptions: VertxEndpointOptions): DecodeInputsResult =
+    decodeBody(DecodeInputs(e.input, new VertxDecodeInputsContext(rc)), rc)
+
+  private def decodeBody(result: DecodeInputsResult, rc: RoutingContext)
+                        (implicit serverOptions: VertxEndpointOptions): DecodeInputsResult = {
+    result match {
+      case values: DecodeInputsResult.Values =>
+        values.bodyInput match {
+          case None => values
+          case Some(bodyInput@EndpointIO.Body(bodyType, codec, _)) =>
+            codec.decode(extractRawBody(bodyType, rc)) match {
+              case DecodeResult.Value(body) => values.setBodyInputValue(body)
+              case failure: DecodeResult.Failure => DecodeInputsResult.Failure(bodyInput, failure): DecodeInputsResult
+            }
+        }
+      case failure: DecodeInputsResult.Failure => failure
+    }
+  }
+
+  private def extractRawBody[B](bodyType: RawBodyType[B], rc: RoutingContext)
+                       (implicit serverOptions: VertxEndpointOptions): Any =
+    bodyType match {
+      case RawBodyType.StringBody(defaultCharset) => rc.getBodyAsString(defaultCharset.toString).get
+      case RawBodyType.ByteArrayBody => rc.getBodyAsString.get.getBytes()
```

Comment:
should we convert the body first to a string,then to bytes? Maybe it would be bettere to read the bytes directly as in the input stream case?

Reply:
True.
Fixed.